File: flutter/lib/payment/mock_payment_provider.dart
import 'payment_models.dart';
import 'payment_provider.dart';

class MockPaymentProvider implements PaymentProvider {
  /// When true, the next/ongoing purchases produce a fail receipt the server
  /// rejects (verifies as a failed order). This is the sandbox failure mode.
  bool failPurchases = false;
  final Set<String> _owned = {};

  @override
  Future<List<StoreProduct>> loadProducts(StoreProductMapping? mapping) async {
    if (mapping == null) return const [];
    return [
      if (mapping.apple != null) StoreProduct(storeProductId: mapping.apple!),
      if (mapping.google != null) StoreProduct(storeProductId: mapping.google!),
      if (mapping.hms != null) StoreProduct(storeProductId: mapping.hms!),
    ];
  }

  @override
  Future<PurchaseResult> purchase(String storeProductId) async {
    final receipt = <String, Object?>{'productId': storeProductId, if (failPurchases) 'fail': true};
    if (!failPurchases) _owned.add(storeProductId);
    return PurchaseResult(storeProductId: storeProductId, receipt: receipt);
  }

  @override
  Future<List<PurchaseResult>> restore() async {
    return _owned
        .map((id) => PurchaseResult(storeProductId: id, receipt: <String, Object?>{'productId': id}))
        .toList(growable: false);
  }
}

File: react-native/src/payment/mockPaymentProvider.ts
import type { PurchaseResult, StoreProduct, StoreProductMapping } from './paymentModels';
import type { PaymentProvider } from './paymentProvider';

export class MockPaymentProvider implements PaymentProvider {
  failPurchases = false;
  private readonly owned = new Set<string>();

  async loadProducts(mapping: StoreProductMapping | null): Promise<readonly StoreProduct[]> {
    if (!mapping) return [];
    return [
      ...(mapping.apple ? [{ storeProductId: mapping.apple }] : []),
      ...(mapping.google ? [{ storeProductId: mapping.google }] : []),
      ...(mapping.hms ? [{ storeProductId: mapping.hms }] : []),
    ];
  }

  async purchase(storeProductId: string): Promise<PurchaseResult> {
    const receipt: Record<string, unknown> = { productId: storeProductId };
    if (this.failPurchases) receipt.fail = true;
    else this.owned.add(storeProductId);
    return { storeProductId, receipt };
  }

  async restore(): Promise<readonly PurchaseResult[]> {
    return [...this.owned].map((id) => ({
      storeProductId: id,
      receipt: { productId: id },
    }));
  }
}
